### ISSUE TITLE
Kernel Up & fully functioning?

### ISSUE THREAD

**Issue by Jasoroony** [varghes/Raspberry-QNX#5]
I'm having trouble using interrupts, I've ported the QNX SD card driver source over to the Pi (and it works) but the only interrupt that seems to fire is #33 (video).  I've attached to that because it reliably (and unfortunately often) calls the handler where I check the SDHOST device registers for what I need.  It never fires int #56 which is what I should be using to be optimal.

I notice that even the included devc-serdebug driver (although it attaches to int #57) never gets a int signal either.

Also, that serial port driver stops working (accepting input) when I run the SD card driver.  I'm not sure why, the SD driver source works on my other (non-Pi) QNX system.  I just don't start it and instead use the default/boot connection (ksh complains that /dev/tty isn't there and there won't be full job control) but I can at least still enter commands and get output, even though it's a bit funky.

I guess my question is, is the kernel really fully functioning?

**Comment by Jasoroony** [comment]
More debugging shows the kernel is in fact enabling the int controller bits correctly (0x2000B200).  Further still, setting BUSY_CMD_SET on SD commands makes the controller raise the interrupt correctly too, so almost everything is working now...

devc-serdebug still stops accepting data from my terminal, but only after I perform a directory listing of the newly mounted SD card (which is weird, because it's already done quite a lot of SD card commands, including reading a few sectors - just not as many as it does reading the root folder).

Thanks for making this available.

**Comment by Jasoroony** [comment]
Hmmm, if I start the devc device before the SD driver, it locks up right away.

**Comment by Jasoroony** [comment]
I must have assumed you connected to int #57 but I can see in the config (startup script) that it's being set to #1 (a timer).

I don't think this is compatible with libcam.so...  Probably a lot of things, actually.

**Comment by Jasoroony** [comment]
Ha!  I did see it attach to # 57, it's hard-coded deep in the code.

What's happening (for future readers)...  devc-serdebug attaches to int 57, this initially works well.  The SD driver then attaches to int 56 which also works, but for some reason masks 57 preventing any more serial input (this is not code in the SD driver source, maybe it's a bug in procnto or maybe libcam.so, who knows).

I now know this is what's happening, because I also wrote a standalone program that sleeps for 30 seconds, wakes, calls InterruptMask(57,-1); and InteruptUnmask(57,-1); and loops to sleep again...  After 30 seconds once booted, the serial port becomes active again suddenly processing the input in the fifo, but the next time I go to access the SD card it doesn't work because its int 56 is now masked so SD command completion is never processed.

For a bit of fun, I modified the serial driver to connect to int 33 (the one that goes off by itself at a rapid pace) and it worked very well too until... I started the SD driver, and the video int also stopped firing (masked) and I'm sure the SD driver source isn't doing blanket mask calls over the full range of interrupts!  So it seems this kernel might have significant issues.

Oh well.

**Comment by Jasoroony** [comment]
LOL - I modified both programs to attach to int # 33 - both devices are now working "properly".  SD card is much slower than it was on the correct interrupt.

A real one int wonder! :)

**Comment by Jasoroony** [comment]
Got the USB mouse and keyboard driver working...  Had to alter the graphics driver too, I'm using adobe/macromedia flash-gf (from an old logitech remote-control firmware download) to make the graphics driver not report success on every (and all) surface requests - only the first one - which makes the GLES gears look better too (if it thinks it has a couple of frame buffers, it tries to use them and it looks messy, but a NULL will make it use software buffers and it works better).  Have had to fix a crash when a surface is freed too, gfi seems to have a bug in there somewhere...

No wonder my (Ford 2011 car's) QNX system didn't implement the "gfi_" stuff.  flash-gf also doesn't translate the full keyboard key-set, they missed a bunch of keys - not so important I guess...

LOL - this stuff is soo cool, and yet so amateurish...  No wonder Ford went to Microsoft and then back again to QNX...  When WinCE had just as many issues......  FYI - I'm not releasing any of this until the kernel's interrupt stuff is working properly here.

**Comment by Jasoroony** [comment]
Ah, nasty, interrupts have been set to max 64 - and I need #66 to implement the sound driver...

I give up.
